x/fees/keeper/msg_server.go

```diff
@@ -0,0 +1,155 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/tharsis/evmos/v3/x/fees/types"
+)
+
+var _ types.MsgServer = &Keeper{}
+
+// RegisterDevFeeInfo registers a contract to receive transaction fees
+func (k Keeper) RegisterDevFeeInfo(
+	goCtx context.Context,
+	msg *types.MsgRegisterDevFeeInfo,
+) (*types.MsgRegisterDevFeeInfoResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if !k.isEnabled(ctx) {
+		return nil, sdkerrors.Wrapf(types.ErrInternalFee, "fees module is not enabled")
+	}
+
+	contract := common.HexToAddress(msg.ContractAddress)
+	if k.IsFeeRegistered(ctx, contract) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "contract %s is already registered", contract)
+	}
+
+	var withdrawal sdk.AccAddress
+	deployer, _ := sdk.AccAddressFromBech32(msg.DeployerAddress)
+	derivedContractAddr := common.BytesToAddress(deployer)
+
+	if msg.WithdrawAddress != "" {
+		withdrawal, _ = sdk.AccAddressFromBech32(msg.WithdrawAddress)
+	}
+
+	// the contract can be directly deployed by an EOA or created through one
+	// or more factory contracts. If it was deployed by an EOA account, then
+	// msg.Nonces contains the EOA nonce for the deployment transaction.
+	// If it was deployed by one or more factories, msg.Nonces contains the EOA
+	// nonce for the origin factory contract, then the nonce of the factory
+	// for the creation of the next factory/contract.
+	for _, nonce := range msg.Nonces {
+		derivedContractAddr = crypto.CreateAddress(derivedContractAddr, nonce)
+	}
+
+	if contract != derivedContractAddr {
+		return nil, sdkerrors.Wrapf(
+			sdkerrors.ErrorInvalidSigner,
+			"not contract deployer or wrong nonce: expected %s instead of %s", derivedContractAddr,
+			contract,
+		)
+	}
+
+	// contract must already be deployed, to avoid spam registrations
+	contractAccount := k.evmKeeper.GetAccountWithoutBalance(ctx, contract)
+	if contractAccount == nil || !contractAccount.IsContract() {
```

Comment:
we should prob check the same for the deployer address (i.e that the address is not a contract)